health/health_virtual.c

```diff
@@ -0,0 +1,168 @@
+                  rrdhost_hostname(host), rrdcalc_chart_name(rcv), rrdcalc_name(rcv),
+                  rcv->calculation->parsed_as, buffer_tostring(rcv->calculation->error_msg)
+                  );
+            buffer_json_member_add_string(wb, "calc_error", buffer_tostring(rcv->calculation->error_msg));
+        } else {
+            rcv->run_flags &= ~RRDCALC_FLAG_CALC_ERROR;
+
+            netdata_log_debug(D_HEALTH, "Health (virtual) on host '%s', alarm '%s.%s': expression '%s' gave value "
+                  NETDATA_DOUBLE_FORMAT
+                  ": %s (source: %s)", rrdhost_hostname(host), rrdcalc_chart_name(rcv), rrdcalc_name(rcv),
+                  rcv->calculation->parsed_as, rcv->calculation->result,
+                  buffer_tostring(rcv->calculation->error_msg), rrdcalc_source(rcv)
+                  );
+
+            rcv->value = rcv->calculation->result;
+            buffer_json_member_add_double(wb, "calc", rcv->value);
+        }
+    }
+
+    if (likely(rcv->warning)) {
+        rcv->warning->value_at = at;
+
+        if (unlikely(!expression_evaluate(rcv->warning))) {
+            // calculation failed
+            rcv->run_flags |= RRDCALC_FLAG_WARN_ERROR;
+        } else {
+            rcv->run_flags &= ~RRDCALC_FLAG_WARN_ERROR;
+            buffer_json_member_add_double(wb, "warn", rcv->warning->result);
+            if (rcv->warning->result) {
+                rcv->status = RRDCALC_STATUS_WARNING;
+                vraised_warn = true;
+            }
+        }
+    }
+
+    if (likely(rcv->critical)) {
+        rcv->critical->value_at = at;
+
+        if (unlikely(!expression_evaluate(rcv->critical))) {
+            // calculation failed
+            rcv->run_flags |= RRDCALC_FLAG_CRIT_ERROR;
+        } else {
+            rcv->run_flags &= ~RRDCALC_FLAG_CRIT_ERROR;
+            buffer_json_member_add_double(wb, "crit", rcv->critical->result);
+            if (rcv->critical->result) {
+                rcv->status = RRDCALC_STATUS_CRITICAL;
+                vraised_crit = true;
+            }
+        }
+    }
+
+    if (!vraised_warn && !vraised_crit)
+        rcv->status = RRDCALC_STATUS_CLEAR;
+
+    buffer_json_object_close(wb);
+}
+
+void health_virtual(RRDHOST *host, BUFFER *wb, struct health_virtual *hv, int min_run_every) {
+
+    buffer_json_member_add_object(wb, "configuration"                  );
+
+    DICTIONARY *dict_rcvs = NULL;
+    dict_rcvs = dictionary_create(DICT_OPTION_SINGLE_THREADED | DICT_OPTION_VALUE_LINK_DONT_CLONE);
+
+    health_config_setup_rc_from_api(wb, host, dict_rcvs, hv);
+    buffer_json_object_close(wb);
+
+    RRDCALC *rcv;
+    dfe_start_read(dict_rcvs, rcv) {
+        buffer_json_member_add_array(wb, string2str(rcv->chart));
+
+        //small adjustment to match health's results
+        time_t now      = now_realtime_sec();
+        time_t at       = hv->after  ? hv->after - 1  : now;
+        time_t before_t = hv->before ? hv->before - 1 : now;
```

Comment:
Considering that suffix `_t` is reserved to data [types](https://www.gnu.org/software/libc/manual/html_node/Reserved-Names.html) like `time_t`, I suggest renaming the variable `before_t`.